File: Array/MinInRotatedSortedArray.py
"""
Find Minimum in Rotated Sorted Array
"""

def findMin(nums) -> int:
    low = 0
    high = len(nums) - 1

    while low < high:
        mid = (high + low)//2

        if nums[mid] == nums[high]:
            high -= 1
        elif nums[mid] > nums[high]:
            low = mid + 1
        else:
            high = mid

    return nums[high]



if __name__ == '__main__':
    print(findMin([8,9,5,6,7]))
    print(findMin([5,6,7,8,9]))
    print(findMin([6,7,8,9,5]))
    print(findMin([16,7]))
    print(findMin([6,6,6,6,6,11]))

File: Array/MinInRotatedSortedArray.cpp
#include <vector>
#include <string>
#include <stack>
#include <unordered_map>
#include <set>
#include <iostream>
#include <algorithm>
#include <iterator>
#include <unordered_set>

using namespace std;

int worker(vector<int> nums, int start, int end) {
    
    
    if (start > end) return -1;
    if (start == end) return nums[start];
    if (end - start == 1) return min(nums[start], nums[end]);
    
    int mid = (end + start) / 2;
    
    if (nums[start] <= nums[mid]) {
        //First part is sorted
        if (nums[end] > nums[start]) {
            return nums[start];
        } else {
            return worker(nums, mid, end);
        }
    } else {
        //Second part is sorted
        if ( nums[start] < nums[end]) {
            return nums[start];
        } else {
            return worker(nums, start, mid);
        }
    }
}

int findMin(vector<int>& nums) {
    return worker(nums, 0, nums.size()-1);
}

int main() {

    vector<int> data{1,2,3,4,5};
    copy(data.begin(), data.end(), ostream_iterator<int>(cout, " "));
    cout<<"Min :"<<findMin(data)<<endl;

    vector<int> data1{9, 1,2,3,4,5, 6, 7, 8};
    copy(data1.begin(), data1.end(), ostream_iterator<int>(cout, " "));
    cout<<"Min :"<<findMin(data1)<<endl;

    vector<int> data2{9,10,11,12, 13, 14, 15,16,17, 1,2,3,4,5, 6, 7, 8};
     copy(data2.begin(), data2.end(), ostream_iterator<int>(cout, " "));
    cout<<"Min :"<<findMin(data2)<<endl;

}